ahnlich/ai/src/manager/mod.rs

```diff
@@ -167,27 +171,69 @@ impl Task for ModelThread {
 
 #[derive(Debug)]
 pub struct ModelManager {
-    models: HashMap<SupportedModels, mpsc::Sender<ModelThreadRequest>>,
+    models: Cache<SupportedModels, mpsc::Sender<ModelThreadRequest>>,
+    supported_models: Vec<SupportedModels>,
+    task_manager: Arc<TaskManager>,
+    // For test purposes
+    has_recreated_models: Arc<AtomicBool>,
 }
 
 impl ModelManager {
     pub async fn new(
         supported_models: &[SupportedModels],
-        task_manager: &TaskManager,
+        task_manager: Arc<TaskManager>,
+        time_to_idle: u64,
     ) -> Result<Self, AIProxyError> {
         // TODO: Actually load the model at this point, with supported models and throw up errors,
         // also start up the various models with the task manager passed in
-        //
-        let mut models = HashMap::with_capacity(supported_models.len());
+        let models = Cache::builder()
+            .max_capacity(supported_models.len() as u64)
+            .time_to_idle(Duration::from_secs(time_to_idle))
+            .build();
         for model in supported_models {
-            // QUESTION? Should we hard code the buffer size or add it to config?
-            let (request_sender, request_receiver) = mpsc::channel(100);
-            // There may be other things needed to load a model thread
-            let model_thread = ModelThread::new(*model, request_receiver);
-            task_manager.spawn_task_loop(model_thread).await;
-            models.insert(*model, request_sender);
+            let request_sender = Self::create_request_sender(model, &task_manager).await?;
+            models.insert(*model, request_sender).await;
         }
-        Ok(Self { models })
+        Ok(Self {
+            models,
+            supported_models: supported_models.to_vec(),
+            task_manager,
+            has_recreated_models: Arc::new(AtomicBool::new(false)),
+        })
+    }
+
+    // QUESTION? Should we hard code the buffer size or add it to config?
+    #[tracing::instrument]
+    async fn create_request_sender(
+        supported_model: &SupportedModels,
+        task_manager: &TaskManager,
```

Comment:
You can unify this to just take self if you create an empty cache in ModelManager::new() and then proceed to call try_get_with for all of your supported models